benchmark/solver/batch_solver.cpp

```diff
@@ -0,0 +1,646 @@
+                static_cast<gko::remove_complex<etype>>(FLAGS_rel_res_goal))
+            .with_preconditioner(gko::preconditioner::batch::type::jacobi)
+            .with_relaxation_factor(static_cast<gko::remove_complex<etype>>(
+                FLAGS_relaxation_factor))
+            .on(exec);
+    }
+    throw std::range_error(std::string("The provided string <") + description +
+                           "> does not match any solver!");
+}
+
+
+void solve_system(const std::string &sol_name,
+                  std::shared_ptr<gko::Executor> exec,
+                  std::shared_ptr<const gko::BatchLinOp> system_matrix,
+                  const batch_vec<etype> *b,
+                  const batch_vec<etype> *scaling_vec, batch_vec<etype> *x,
+                  rapidjson::Value &test_case,
+                  rapidjson::MemoryPoolAllocator<> &allocator)
+{
+    try {
+        auto &solver_case = test_case["batch_solver"];
+        auto solver_name = sol_name.c_str();
+        if (!FLAGS_overwrite && solver_case.HasMember(solver_name)) {
+            return;
+        }
+
+        add_or_set_member(solver_case, solver_name,
+                          rapidjson::Value(rapidjson::kObjectType), allocator);
+        auto &solver_json = solver_case[solver_name];
+        add_or_set_member(solver_json, "scaling",
+                          rapidjson::StringRef(FLAGS_batch_scaling.c_str()),
+                          allocator);
+        size_type nbatch = system_matrix->get_num_batch_entries();
+        if (FLAGS_detailed && b->get_size().at(0)[1] == 1 && !FLAGS_overhead) {
+            add_or_set_member(solver_json, "rhs_norm",
+                              rapidjson::Value(rapidjson::kObjectType),
+                              allocator);
+            for (size_type i = 0; i < nbatch; ++i) {
+                add_or_set_member(
+                    solver_json["rhs_norm"], std::to_string(i).c_str(),
+                    rapidjson::Value(rapidjson::kArrayType), allocator);
+                auto unbatch_b = b->unbatch();
+                solver_json["rhs_norm"][std::to_string(i).c_str()].PushBack(
+                    compute_norm2(lend(unbatch_b[i])), allocator);
+            }
+        }
+
+        for (auto stage : {"generate", "apply"}) {
+            add_or_set_member(solver_json, stage,
+                              rapidjson::Value(rapidjson::kObjectType),
+                              allocator);
+            add_or_set_member(solver_json[stage], "components",
+                              rapidjson::Value(rapidjson::kObjectType),
+                              allocator);
+        }
+
+        std::shared_ptr<gko::log::BatchConvergence<etype>> logger =
+            gko::log::BatchConvergence<etype>::create(exec);
+        // warm run
+        for (unsigned int i = 0; i < FLAGS_warmup; i++) {
+            auto x_clone = clone(x);
+            auto solver =
+                generate_solver(exec, sol_name)->generate(system_matrix);
```

Comment:
This is a problem, I think. You need to create a copy of the system matrix here, because with scaling, it will be modified by the solver. Then all the further solves will be wrong.

Reply:
Good catch. Should be fixed now.